incremental-font-transfer/src/patch_group.rs

```diff
@@ -0,0 +1,1253 @@
+};
+
+/// A group of patches derived from a single IFT font which can be applied simultaneously
+/// to that font. Patches are initially missing data which must be fetched and supplied to
+/// the group before it can be applied to the font.
+pub struct PatchGroup<'a> {
+    font: FontRef<'a>,
+    patches: Option<CompatibleGroup>,
+}
+
+impl<'a> PatchGroup<'a> {
+    /// Intersect the available and unapplied patches in ift_font against subset_definition and return a group
+    /// of patches which would be applied next.
+    pub fn select_next_patches<'b>(
+        ift_font: FontRef<'b>,
+        subset_definition: &SubsetDefinition,
+    ) -> Result<PatchGroup<'b>, ReadError> {
+        let candidates = intersecting_patches(&ift_font, subset_definition)?;
+        if candidates.is_empty() {
+            return Ok(PatchGroup {
+                font: ift_font,
+                patches: None,
+            });
+        }
+
+        let compat_group = Self::select_next_patches_from_candidates(
+            candidates,
+            ift_font.ift().ok().map(|t| t.compatibility_id()),
+            ift_font.iftx().ok().map(|t| t.compatibility_id()),
+        )?;
+
+        Ok(PatchGroup {
+            font: ift_font,
+            patches: Some(compat_group),
+        })
+    }
+
+    /// Returns the list of URIs in this group.
+    pub fn uris(&self) -> impl Iterator<Item = &str> {
+        self.invalidating_patch_iter()
+            .chain(self.non_invalidating_patch_iter())
+            .map(|info| info.uri.as_str())
+    }
+
+    pub fn has_uris(&self) -> bool {
+        let Some(patches) = &self.patches else {
+            return false;
+        };
+        match patches {
+            CompatibleGroup::Full(FullInvalidationPatch(_)) => true,
+            CompatibleGroup::Mixed { ift, iftx } => ift.has_uris() || iftx.has_uris(),
+        }
+    }
+
+    fn next_invalidating_patch(&self) -> Option<&PatchInfo> {
+        self.invalidating_patch_iter().next()
+    }
+
+    fn invalidating_patch_iter(&self) -> impl Iterator<Item = &PatchInfo> {
+        let full = match &self.patches {
+            Some(CompatibleGroup::Full(info)) => Some(&info.0),
+            _ => None,
+        };
+
+        let partial_1 = match &self.patches {
+            Some(CompatibleGroup::Mixed {
+                ift: ScopedGroup::PartialInvalidation(v),
+                iftx: _,
+            }) => Some(&v.0),
+            _ => None,
+        };
+
+        let partial_2 = match &self.patches {
+            Some(CompatibleGroup::Mixed {
+                ift: _,
+                iftx: ScopedGroup::PartialInvalidation(v),
+            }) => Some(&v.0),
+            _ => None,
+        };
+
+        full.into_iter().chain(partial_1).chain(partial_2)
+    }
+
+    fn non_invalidating_patch_iter(&self) -> impl Iterator<Item = &PatchInfo> {
+        let ift = match &self.patches {
+            Some(CompatibleGroup::Mixed { ift, iftx: _ }) => Some(ift),
```

Comment:
ditto here, it could just be,
```suggestion
            Some(CompatibleGroup::Mixed { ift, .. }) => Some(ift),
```